docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,369 @@
+        message: <human readable message indicating details about the transition>
+```
+
+> NOTE: The above custom resource YAML serves as a template which will be used to further create a CRD and Golang APIs.
+
+### Golang API
+
+```go
+// EtcdOperatorTask represents an out-of-band operator task.
+type EtcdOperatorTask struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // Spec is the specification of the task.
+  Spec EtcdOperatorTaskSpec
+  // Status is most recently observed status of the task.
+  Status EtcdOperatorTaskStatus
+}
+```
+
+#### Spec
+
+The authors propose that the following fields should be specified in the spec (desired state) of the `EtcdOperatorTask` custom resource.
+
+* To capture the type of `out-of-band` operator task to be performed, `spec.Type` field should be defined. It can have values from any supported "out-of-band" operations/tasks eg. "OnDemandSnaphotTask", "QuorumLossRecoveryTask" etc.
+* To capture the configuration specific to each task, a `spec.Config` field is defined of type [RawExtension](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L49-L102) as each task can have different input configuration.
+
+```go
+// EtcdOperatorTaskSpec is the spec for a EtcdOperatorTask resource.
+type EtcdOperatorTaskSpec struct {
+  
+  // Type specifies the type of out-of-band operator task to be performed. 
+  Type string `json:"taskType"`
+
+  // Config is a task specific configuration.
+  Config *runtime.RawExtension `json:"taskConfig,omitempty"`
+
+  // TTLSecondsAfterFinished is the time-to-live after which the task and 
+  // related resources will be garbage collected.
+  // +optional
+  TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+```
+
+#### Status
+
+The authors propose that the following fields should be specified in the `Status` (current state) of the EtcdOperatorTask custom resource as the custom resource's `Status` will be used to monitor the progress of the task.
+
+* To capture the Status of a task, a field `taskStatus` is defined in status.
+* If operation involves many stages, so to capture the status of intermediate or any stage, `.status.lastOperation` will be useful.
+
+```go
+// TaskState represents the state of the task.
+type TaskState string
+
+const (
+  Failed TaskState = "failed"
+  Pending TaskState = "pending"
+  Rejected TaskState = "rejected"
+  Completed TaskState = "completed"
+  InProgress TaskState = "inProgress"
+)
+
+// EtcdOperatorTaskStatus is the status for a EtcdOperatorTask resource.
+type EtcdOperatorTaskStatus struct {
+  // ObservedGeneration is the most recent generation observed for the resource.
+  ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+  // State of the task is the last known state of the task.
+  State TaskState `json:"taskStatus"`
+  // Time at which operation has been triggered.
+  InitiatedAt metav1.Time `json:"initiatedAt"`
```

Comment:
Same as above. Also, What exactly does `Time at which operation has been triggered` mean? Is it the same as creationTimestamp of the CR? Or is it the time at which druid picked up the resource for actuation? Or something else entirely? This seems ambiguous to me right now.